### PR TITLE
Fix bug in dense_dot_range

### PR DESCRIPTION
**Pull request by theartful**
This fixes a buffer underflow bug in `dense_dot_range`.

### CODE REVIEW

File: src/shogun/features/DotFeatures.cpp [hunk 1]
```diff
@@ -51,9 +51,6 @@ float64_t CDotFeatures::dense_dot_sgvec(int32_t vec_idx1, SGVector<float64_t> ve
 void CDotFeatures::dense_dot_range(float64_t* output, int32_t start, int32_t stop, float64_t* alphas, float64_t* vec, int32_t dim, float64_t b) const
 {
 	ASSERT(output)
-	// write access is internally between output[start..stop] so the following
-	// line is necessary to write to output[0...(stop-start-1)]
-	output-=start;
```
Comment (theartful) [inline src/shogun/features/DotFeatures.cpp]: output was assumed to be accessed from `start` to `stop`, which wasn't the case.

File: src/shogun/features/DotFeatures.cpp [hunk 2]
```diff
@@ -81,7 +78,7 @@ void CDotFeatures::dense_dot_range(float64_t* output, int32_t start, int32_t sto
 #endif
 
 		int32_t t_start=thread_num*step;
```
Comment (theartful) [inline src/shogun/features/DotFeatures.cpp]: `t_start` starts from 0

File: src/shogun/features/DotFeatures.cpp [hunk 3]
```diff
@@ -93,9 +90,9 @@ void CDotFeatures::dense_dot_range(float64_t* output, int32_t start, int32_t sto
 #endif
 		{
 			if (alphas)
-				output[i]=alphas[i]*this->dense_dot(i, vec, dim)+b;
```
Comment (theartful) [inline src/shogun/features/DotFeatures.cpp]: and here happens buffer underflow if `t_start != 0`

File: src/shogun/features/DotFeatures.cpp [hunk 4]
```diff
@@ -81,7 +78,7 @@ void CDotFeatures::dense_dot_range(float64_t* output, int32_t start, int32_t sto
 #endif
 
 		int32_t t_start=thread_num*step;
-		int32_t t_stop=(thread_num==num_threads) ? stop : (thread_num+1)*step;
+		int32_t t_stop=(thread_num==num_threads) ? num_vectors : (thread_num+1)*step;
 
 #ifdef WIN32
 		for (int32_t i=t_start; i<t_stop; i++)
```
Comment (iglesias) [inline src/shogun/features/DotFeatures.cpp]: Can the TODO below be removed?

File: src/shogun/features/DotFeatures.cpp [hunk 5]
```diff
@@ -93,9 +90,9 @@ void CDotFeatures::dense_dot_range(float64_t* output, int32_t start, int32_t sto
 #endif
 		{
 			if (alphas)
-				output[i]=alphas[i]*this->dense_dot(i, vec, dim)+b;
+				output[i]=alphas[i]*this->dense_dot(i + start, vec, dim)+b;
```
Comment (iglesias) [inline src/shogun/features/DotFeatures.cpp]: I think `start` is included earlier.

To make sure everything is fine, have you tested if with different number of threads?

File: tests/unit/features/DotFeatures_unittest.cc [hunk 6]
```diff
@@ -121,3 +121,47 @@ TEST_F(DotFeaturesTest, compute_cov_nocopy)
 	for (index_t i = 0; i < (index_t)cov.size(); ++i)
 		EXPECT_NEAR(cov[i], ref_cov_ab[i], eps);
 }
+
+TEST_F(DotFeaturesTest, dense_dot_range)
+{
```
Comment (iglesias) [inline tests/unit/features/DotFeatures_unittest.cc]: Thanks for including the test!
What about including also a much more trivial one, whose expected result could be seen just from reading the test code in a second?
Also, I think it would be good to ensure test coverage with different number of threads. Is that currently happening?

File: src/shogun/features/DotFeatures.cpp (hunk 5 above)
Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: `i` ranges from `t_start` to `t_stop`. But `t_start` begins from `0` and `t_stop` goes up to `stop`.
It should be that `t_start` begins from `0` and `t_stop` goes up to `stop - start`, or `t_start` begins from `start` and `t_stop` goes up to `stop`.

Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: Judging from this comment
``` C++
// write access is internally between output[start..stop] so the following
// line is necessary to write to output[0...(stop-start-1)]
output-=start;
```
This was initially written so that we go from `start` to  `end`, so we subtract `start` from the output, so it transforms to `0` to `end - start`.

Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: This assumption didn't hold, as `i` ranges from `0` to `t_stop`
```C++
int32_t t_start=thread_num*step;
int32_t t_stop=(thread_num==num_threads) ? stop : (thread_num+1)*step;
```
And even if it did, the indexing of `alpha` would be wrong.

File: tests/unit/features/DotFeatures_unittest.cc (hunk 6 above)
Comment (theartful) [reply tests/unit/features/DotFeatures_unittest.cc]: I believe it should run concurrently on any machine with `openmp`

File: src/shogun/features/DotFeatures.cpp (hunk 4 above)
Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: Should we also remove the progress bar? I don't think such an operation needs it.

File: src/shogun/features/DotFeatures.cpp (hunk 5 above)
Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: Should I revert the changes and write a simple test that fails?

File: src/shogun/features/DotFeatures.cpp (hunk 4 above)
Comment (iglesias) [reply src/shogun/features/DotFeatures.cpp]: Why not? It could be useful if many vectors dotted.

File: src/shogun/features/DotFeatures.cpp (hunk 5 above)
Comment (iglesias) [reply src/shogun/features/DotFeatures.cpp]: That sounds like a good idea. You can add it to another branch if that's more efficient than reverting. Link to test results or copy+paste suffices if that is faster as well.

`t_start` starts on a different value depending on the thread.

File: tests/unit/features/DotFeatures_unittest.cc (hunk 6 above)
Comment (iglesias) [reply tests/unit/features/DotFeatures_unittest.cc]: Let me rephrase, is there some test exercising this code with more than one thread?
Alternatively, have you tested it locally with different number of threads (at least one case with more than one).

File: src/shogun/features/DotFeatures.cpp (hunk 5 above)
Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: I ran this test on the old version, which tests the underflow:
```C++
TEST_F(DotFeaturesTest, dense_dot_range)
{
	index_t num_feats = 2;
	index_t num_vectors = 4;

	SGMatrix<float64_t> data(num_feats, num_vectors);
	for (index_t i = 0; i < num_vectors; i++)
		data.get_column(i).set_const(5);
	auto feats = some<CDenseFeatures<float64_t>>(data);

	SGVector<float64_t> vec(num_feats);
	vec.set_const(3);

	index_t start = 1;
	index_t stop = num_vectors;
	
	float64_t* buffer = new float64_t[num_vectors];
	float64_t& gaurd = buffer[0];
	gaurd = 0;
	float64_t* output = buffer  + 1;

	feats->dense_dot_range(
	    output, start, stop, nullptr, vec.vector, num_feats, 0);

	ASSERT_EQ(gaurd, 0);
}
```
Here `gaurd` should not be changed at all.
These are the results:
```
$ ./shogun-unit-test --gtest_filter=DotFeaturesTest.dense_dot_range
Note: Google Test filter = DotFeaturesTest.dense_dot_range
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DotFeaturesTest
[ RUN      ] DotFeaturesTest.dense_dot_range
/home/theartful/Programming/shogun/tests/unit/features/DotFeatures_unittest.cc:149: Failure
      Expected: gaurd
      Which is: 30
To be equal to: 0
[  FAILED  ] DotFeaturesTest.dense_dot_range (12 ms)
[----------] 1 test from DotFeaturesTest (12 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (15 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DotFeaturesTest.dense_dot_range

 1 FAILED TEST
```

File: tests/unit/features/DotFeatures_unittest.cc (hunk 6 above)
Comment (theartful) [reply tests/unit/features/DotFeatures_unittest.cc]: Locally `openmp` spawns 4 threads on my machine to do the operation.

File: src/shogun/features/DotFeatures.cpp (hunk 4 above)
Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: But does the user need to see them? I think the progress bar would make sense if it indicated the state of the algorithm, instead of an operation.

Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: About the TODO. Do you mean remove it with the signal stuff, or implement it? If so, I would need to first check out what are these signals about :"D

File: src/shogun/features/DotFeatures.cpp (hunk 5 above)
Comment (iglesias) [reply src/shogun/features/DotFeatures.cpp]: Memory leak with `buffer`?
In any case, use std::vector instead.

Comment (theartful) [reply src/shogun/features/DotFeatures.cpp]: The test leaks indeed. Freeing the memory doesn't change the test result. The problem is that the old `dense_dot_range` changes `output[-1]`

Comment (iglesias) [reply src/shogun/features/DotFeatures.cpp]: All right, the point is clear now. Thanks for writing this test, good illustration.
